### PR TITLE
Implemented split routines and a test

### PR DESCRIPTION
**Pull request by sbak5**
- split, vsplit, hsplit, dsplit and array_split
    - test code to run and compare 'split' and 'array_split' w/ cunumeric and numpy

### CODE REVIEW

**Comment by sbak5** [comment]
The test code can run 'split' and 'array_split' with different input configurations. (-b <min number of elements in each dimension for 3D tensor>, -e <max # elems>
The following error happens randomly. try to run w/ the following argument
command: legate --cpus 16 --gpus 1 --fbmem 15000 /workspace/nfs/cunumeric/tests/split.py -b 90 -e 94"

Testing np.split((92, 90, 91), [111, 108, 105, 102, 99, 96, 93, 90, 87, 84, 81, 78, 75, 72, 69, 66, 63, 60, 57, 54, 51, 48, 45, 42, 39, 36, 33, 30, 27, 24, 21, 18, 15, 12, 9, 6, 3], 1) ->  Passed
// This has been passed. After this test, some runtime codes incurs the following error.

[0 - 7fb0a1a5e000]   23.861729 {5}{cunumeric.mapper}: Mapper cunumeric on Node 0 failed to map region requirement 0 of inline mapping (UID 68682) into Visible to all processors on a node memory 1e00000000000000
[0 - 7fb0a1a5e000]   23.861763 {5}{legate}: Legate called abort in core/mapping/base_mapper.cc at line 941 in function report_failed_mapping
Signal 6 received by node 0, process 64 (thread 7fb0a1a5e000) - obtaining backtrace
Signal 6 received by process 64 (thread 7fb0a1a5e000) at: stack trace: 16 frames
  [0] = /usr/lib/x86_64-linux-gnu/libc.so.6(+0x46210) [0x7fb0ab4e3210]
  [1] = /usr/lib/x86_64-linux-gnu/libc.so.6(gsignal+0xcb) [0x7fb0ab4e318b]
  [2] = /usr/lib/x86_64-linux-gnu/libc.so.6(abort+0x12b) [0x7fb0ab4c2859]
  [3] = /workspace/nfs/legate.core/install/lib/liblgcore.so(legate::mapping::BaseMapper::report_failed_mapping(Legion::Mappable const&, unsigned int, Realm::Memory, int)+0x26a) [0x7faf79fbd7ca]
  [4] = /workspace/nfs/legate.core/install/lib/liblgcore.so(legate::mapping::BaseMapper::map_raw_array(Legion::Internal::MappingCallInfo*, Legion::Mappable const&, unsigned int, Legion::LogicalRegion, unsigned int, Realm::Memory, Realm::Processor, std::vector<Legion::Mapping::PhysicalInstance, std::allocator<Legion::Mapping::PhysicalInstance> > const&, Legion::Mapping::PhysicalInstance&, bool, int)+0x11c3) [0x7faf79fbf3a3]
  [5] = /workspace/nfs/legate.core/install/lib/liblgcore.so(legate::mapping::BaseMapper::map_inline(Legion::Internal::MappingCallInfo*, Legion::InlineMapping const&, Legion::Mapping::Mapper::MapInlineInput const&, Legion::Mapping::Mapper::MapInlineOutput&)+0x15c) [0x7faf79fc005c]
  [6] = /workspace/nfs/legate.core/install/lib/liblegion.so(Legion::Internal::MapperManager::invoke_map_inline(Legion::Internal::MapOp*, Legion::Mapping::Mapper::MapInlineInput*, Legion::Mapping::Mapper::MapInlineOutput*, Legion::Internal::MappingCallInfo*)+0x50) [0x7fb0ad427a30]
  [7] = /workspace/nfs/legate.core/install/lib/liblegion.so(Legion::Internal::MapOp::invoke_mapper(Legion::Internal::InstanceSet&, std::vector<Legion::Internal::PhysicalManager*, std::allocator<Legion::Internal::PhysicalManager*> >&)+0x134) [0x7fb0ad01c3d4]
  [8] = /workspace/nfs/legate.core/install/lib/liblegion.so(Legion::Internal::MapOp::trigger_mapping()+0xf2) [0x7fb0ad031582]
  [9] = /workspace/nfs/legate.core/install/lib/liblegion.so(Legion::Internal::Runtime::legion_runtime_task(void const*, unsigned long, void const*, unsigned long, Realm::Processor)+0x706) [0x7fb0ad3f8fc6]
  [10] = /workspace/nfs/legate.core/install/lib/librealm.so(+0x4ba39a) [0x7fb0abd5039a]
  [11] = /workspace/nfs/legate.core/install/lib/librealm.so(+0x2fa113) [0x7fb0abb90113]
  [12] = /workspace/nfs/legate.core/install/lib/librealm.so(+0x2fa1ca) [0x7fb0abb901ca]
  [13] = /workspace/nfs/legate.core/install/lib/librealm.so(+0x2fc8c1) [0x7fb0abb928c1]
  [14] = /workspace/nfs/legate.core/install/lib/librealm.so(+0x2e11ac) [0x7fb0abb771ac]
  [15] = /usr/lib/x86_64-linux-gnu/libc.so.6(+0x5e660) [0x7fb0ab4fb660]

**Comment by magnatelee** [comment]
Looks like you're just running out of the system memory, which you can fix by passing a value greater than 4096 to `--sysmem`. FYI, passing a cuNumeric array to a canonical NumPy call, such as `np.array_equal`, materializes it on a system memory, so you need to make sure the system memory is big enough for your data.

**Comment by sbak5** [comment]
> 

The option resolves the error above. Thanks!

File: cunumeric/module.py [hunk 1]
```diff
@@ -523,6 +523,121 @@ def tile(a, reps):
...
+
+        # the requested # of subarray is larger than the size of array
+        # -> size of 1 subarrays + empty subarrays
+        if len_subarr == 0:
+            len_subarr = 1
+            end_idx = indices
+        else:
+            if res != 0:
+                # The first 'res' groups have len_subarr+1 elements
+                split_pts = [
```
Comment (magnatelee) [inline cunumeric/module.py:579]: can be simplified to list(range(len_subarr + 1, (len_subarr + 1) * res, len_subarr + 1))

File: cunumeric/module.py [hunk 2]
```diff
@@ -523,6 +523,121 @@ def tile(a, reps):
...
+                    i
+                    for i in range(
+                        len_subarr + 1, (len_subarr + 1) * res, len_subarr + 1
+                    )
+                ]
+                first_idx = (len_subarr + 1) * res
+        split_pts += [i for i in range(first_idx, end_idx + 1, len_subarr)]
```
Comment (magnatelee) [inline cunumeric/module.py:586]: I think you better use `.append` or `.extend` for in-place updates. this should be rewritten to `split_pts.extend(range(first_idx, end_idx + 1, len_subarr)))`

File: cunumeric/module.py [hunk 3]
```diff
@@ -523,6 +523,121 @@ def tile(a, reps):
...
+    else:
+        raise ValueError("Integer or array for split should be provided")
+
+    result = []
+    start_idx = 0
+    end_idx = 0
+    out_shape = []
+    in_shape = []
+
+    for i in range(array.ndim):
```
Comment (magnatelee) [inline cunumeric/module.py:598]: the same comment as above

**Comment by manopapad** [comment]
Please don't force-push changes to a branch after it had been added to a PR. This makes it harder for reviewers to track changes.

File: cunumeric/module.py [hunk 4]
```diff
@@ -523,6 +523,112 @@ def tile(a, reps):
...
+        # For a split point, which is larger than the dimension for splitting,
+        # The last non-empty subarray should be copied from
+        # array[last_elem:array.shape[axis]]
+        if pts > array.shape[axis]:
+            end_idx = array.shape[axis]
+        out_shape[axis] = (end_idx - start_idx) + 1
+        in_shape[axis] = slice(start_idx, end_idx)
+        new_subarray = None
+        if start_idx < array.shape[axis] and start_idx < end_idx:
+            new_subarray = array[(tuple)(in_shape)].copy()
```
Comment (manopapad) [inline cunumeric/module.py:616]: do you need the parens around `(tuple)`?

Comment (manopapad) [inline cunumeric/module.py:616]: Don't `copy()` the sub-array, the documentation for `np.split` suggests that it returns views over the base array.

File: cunumeric/module.py [hunk 5]
```diff
@@ -523,6 +523,112 @@ def tile(a, reps):
...
+        if array.ndim >= axis:
+            if equal and array.shape[axis] % indices == 0:
+                raise ValueError(
+                    "array split does not result in an equal divison"
+                )
+        else:
+            raise ValueError(
+                "array({}) has less dimensions than axis({})".format(
+                    array.shape, axis
+                )
```
Comment (manopapad) [inline cunumeric/module.py:568]: I suggest we use f-strings exclusively in new code. We don't have to support python < 3.7.

File: cunumeric/module.py [hunk 6]
```diff
@@ -523,6 +523,112 @@ def tile(a, reps):
...
+    dtype = type(indices)
+    split_pts = []
+
+    if dtype == int:
+        res = array.shape[axis] % indices
+        len_subarr = array.shape[axis] // indices
+        first_idx = len_subarr
+        end_idx = array.shape[axis]
+
+        if array.ndim >= axis:
```
Comment (manopapad) [inline cunumeric/module.py:559]: This should be `array.ndim > axis`

File: cunumeric/module.py [hunk 7]
```diff
@@ -523,6 +523,112 @@ def tile(a, reps):
...
+            if equal and array.shape[axis] % indices == 0:
```
Comment (manopapad) [inline cunumeric/module.py:560]: This should be `!= 0`

File: cunumeric/module.py [hunk 8]
```diff
@@ -523,6 +523,112 @@ def tile(a, reps):
...
+            if res != 0:
+                # The first 'res' groups have len_subarr+1 elements
+                split_pts = list(
+                    range(
+                        len_subarr + 1, (len_subarr + 1) * res, len_subarr + 1
+                    )
+                )
+                first_idx = (len_subarr + 1) * res
+        split_pts.extend(range(first_idx, end_idx + 1, len_subarr))
+    elif dtype == np.array or dtype == list:
```
Comment (manopapad) [inline cunumeric/module.py:586]: The NumPy implementation also supports tuples. I suggest if `indices` is not an integer we just convert it to a list using `list(indices)`. This should handle any kind of sequence.

File: cunumeric/module.py [hunk 9]
```diff
@@ -523,6 +523,112 @@ def tile(a, reps):
...
+                "array({}) has less dimensions than axis({})".format(
+                    array.shape, axis
+                )
+            )
+
+        # the requested # of subarray is larger than the size of array
+        # -> size of 1 subarrays + empty subarrays
+        if len_subarr == 0:
+            len_subarr = 1
+            end_idx = indices
```
Comment (manopapad) [inline cunumeric/module.py:575]: You also need to set `start_idx` to `1`

**Comment by manopapad** [comment]
A general comment, it might be possible to reuse the logic in the original NumPy implementation for this function.

File: cunumeric/module.py [hunk 10]
```diff
@@ -523,6 +523,112 @@ def tile(a, reps):
...
+        new_subarray = None
+        if start_idx < array.shape[axis] and start_idx < end_idx:
+            new_subarray = array[(tuple)(in_shape)].copy()
+        else:
+            out_shape[axis] = 0
+            new_subarray = ndarray(tuple(out_shape), dtype=array.dtype)
+        result.append(new_subarray)
+        start_idx = pts
+
+    # If the last element in `indices` is larger than array.shape[axis],
```
Comment (manopapad) [inline cunumeric/module.py:623]: The behavior of this function for `indices > shape[axis]` is incorrect. For `array_split(np.arange(4), 7)` vanilla NumPy produces
```
[array([0]), array([1]), array([2]), array([3]), array([], dtype=int64), array([], dtype=int64), array([], dtype=int64)]
```
while this function produces
```
[array([0]), array([0]), array([1]), array([2]), array([3]), array([0]), array([0]), array([0]), array([0])]
```

File: cunumeric/module.py [hunk 11]
```diff
@@ -523,6 +523,112 @@ def tile(a, reps):
...
+    # an empty dummy array should be created
+    if start_idx > array.shape[axis]:
```
Comment (manopapad) [inline cunumeric/module.py:625]: This function sometimes skips the last element from the output. For `array_split(np.arange(4), [1,3]))` vanilla NumPy returns
```
[array([0]), array([1, 2]), array([3])]
```
while this function returns
```
[array([0]), array([1, 2])]
```

File: cunumeric/module.py (hunk 6 above)
Comment (sbak5) [reply cunumeric/module.py:559]: axis starts from 0 so when the array is 3D, the maximum valid value for axis is 2.

Comment (manopapad) [reply cunumeric/module.py:559]: Correct. For a 3d array `array.ndim == 3`, so the non-error path should only be taken for `axis < 3`.

Comment (manopapad) [reply cunumeric/module.py:559]: Actually if `axis >= array.ndim`, this code already crashes at the `array.shape[axis]` access on line 554. This check should be moved to the beginning of the function, to also cover the cases where `indices` is not an integer.

Comment (sbak5) [reply cunumeric/module.py:559]: Got it. I'll update the code as you suggested.

File: cunumeric/module.py (hunk 10 above)
Comment (sbak5) [reply cunumeric/module.py:623]: No, the dummy arrays are empty here.

Comment (manopapad) [reply cunumeric/module.py:623]: You're right, I had probably been making changes to the code when I got this output. I reset the code and indeed the dummy arrays are empty. However, it seems like the wrong number is added. For the following example:

```
import numpy as np
import cunumeric as cn
print(np.array_split(np.arange(4), 7))
print(cn.array_split(cn.arange(4), 7))
```

I get the output:

```
[array([0]), array([1]), array([2]), array([3]), array([], dtype=int64), array([], dtype=int64), array([], dtype=int64)]
[array([], dtype=int64), array([0]), array([1]), array([2]), array([3]), array([], dtype=int64), array([], dtype=int64), array([], dtype=int64), array([], dtype=int64)]
```

File: cunumeric/module.py (hunk 4 above)
Comment (sbak5) [reply cunumeric/module.py]: changed to view()

File: cunumeric/module.py (hunk 10 above)
Comment (sbak5) [reply cunumeric/module.py]: Added this case to the test

File: cunumeric/module.py (hunk 11 above)
Comment (sbak5) [reply cunumeric/module.py]: This is also fixed in the new commit.

**Comment by sbak5** [comment]
Removed negative tests in split.py which confirms whether cunumeric and numpy produce the same error for non-viable options w/ split. array_split works for any integer and array arguments so testing array_split w/ diff configurations works for all other routines.

**Comment by manopapad** [comment]
@sbak5 as I asked you before, please don't force-push changes to your branches after you have opened a PR. Having separate commits for fixes helps reviewers do their work.

File: cunumeric/module.py [hunk 12]
```diff
@@ -523,6 +523,105 @@ def tile(a, reps):
...
+            if res != 0:
+                # The first 'res' groups have len_subarr+1 elements
+                split_pts = list(
+                    range(
+                        len_subarr + 1, (len_subarr + 1) * res, len_subarr + 1
+                    )
+                )
+                first_idx = (len_subarr + 1) * res
+        split_pts.extend(range(first_idx, end_idx + 1, len_subarr))
+    elif dtype == np.array or dtype == list or dtype == tuple:
```
Comment (manopapad) [inline cunumeric/module.py:582]: In the case where `indices` is a NumPy array also check that it's an integer array (i.e. `indices.dtype == int`)

Comment (manopapad) [inline cunumeric/module.py:582]: The NumPy type check is wrong; it should be `type(indices) == np.ndarray`, not `type(indices) == np.array` (`np.array` is a function, not a type).

**Comment by manopapad** [comment]
Looks good to me. Please address the last couple of comments above, do a merge from `branch-22.01`, and verify that CI passes.

**Comment by sbak5** [comment]
I'll not do the force-push again. Rebased this branch on the latest commit and resolve the comments above.

**Pull request merged**